# DIVEIN
[DIVEIN](https://divein.fredhutch.org) is a web service that performs automated maximum likelihood phylogenetic analyses of nucleotide and amino acid sequences and other sequence related analyses.

## Usage

Starting with a set of aligned sequences, DIVEIN can:

  - calculate phylogenetic trees via a variety of evolutionary models by three programs of PhyML, FastTree and RAxML.
  - reconstruct the consensus, Most Recent Common Ancestor (MRCA) and Center of Tree (COT) sequences.
  - re-root trees at COT.
  - compute genetic distance distributions, diversity, and divergence from the consensus, MRCA, COT and any sequence in the alignment.
  - compute hamming distance distributions and diversity.
  - graphically represent the inferred tree and plots of divergence, diversity, and distance distribution histograms.
  - detect, visualize and numerically summarize phylogenetically informative sites as well as private mutations.
  - compare intra-individual genetic sequence diversity between populations.
  - define clusters by a genetic similarity cluster identification tool that includes two approaches (TN93 pairwise genetic distance and Phylogenetic branch lengths).

## Citation

Deng W, Maust BS, Nickle DC, Learn GH, Liu Y, Heath L, Kosakovsky Pond SL, Mullins JI 2010 DIVEIN: a web server to analyze phylogenies, sequence divergence, diversity, and informative sites. 
[BioTechniques 48:405-408](https://www.tandfonline.com/doi/full/10.2144/000113370). 
