<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
	<title>DIVEIN</title>
	<link href='static/css/divein.css' media="screen" rel="Stylesheet" type="text/css">
	<script type="text/javascript" src='/static/js/divein.js'></script>
</head>
<body width=100%">	
    <div>
    	<div class="title">DIVEIN</div>
    	<div><img src="/static/images/Jim&Lisa/header.jpg" width='400' height='120' style="float:left"></div>
    	<div class="s-title">Divergence, Diversity,<br>Informative Sites and<br>Phylogenetic Analyses</div>
    </div>
    <div class="spacer"></div>  
    <div class="navbar">
		<a href='index.html' class="nav"><strong>Home</strong></a>
		<a href="insites.html" >Informative sites</a>
		<a href="cot.html" >Center Of Tree</a>
		<div class="dropdown">
			<button class="dropbtn">Phylogeny/Divergence/Diversity 
			  <i class="fa fa-caret-down"></i>
			</button>
			<div class="dropdown-content">
			  <a href="diver.html">PhyML 3.3.20220408</a>
			  <a href="fasttree.html">FastTree 2.1.10</a>
			  <a href="raxml.html">RAxML 8.2.12</a>
			  <a href="hd.html">Hamming distance</a>
			</div>
	  	</div> 
		<a href="tst.html" >Two-Sample Tests</a>
		<a href="cluster.html" >Sequence clustering</a>
		<a href="retrieve.html" >Retrieve results</a>
		<a href='contact.html' >Contact</a>
		<a href='help.html' >Help</a>
  	</div>
    <div id="indent">
    <!-- maintainance message goes here -->
<!--	<div class="alert alert-warning" role="alert">
	         <font color=red><strong>Heads up!</strong>
			            This server will be down for the lab moving from <strong>19 July 2018 to end of July</strong>. 
						Please submitting jobs after then.</font>
	</div>-->
    <!-- end of message -->
	<div>
	<br>
	<p><strong>
	DIVEIN is a web service that performs automated maximum likelihood phylogenetic analyses of nucleotide and amino acid sequences. 
	Starting with a set of aligned sequences, DIVEIN can:</strong>
	<ul>
	<li>calculate phylogenetic trees via a variety of evolutionary models by three programs of PhyML, FastTree and RAxML.
	<li>reconstruct the consensus, Most Recent Common Ancestor (MRCA) and Center of Tree (COT) sequences.
	<li>re-root trees at COT.
	<li>compute genetic distance distributions, diversity, and divergence from the consensus, MRCA, COT and any sequence in the alignment.
	<li>compute hamming distance distributions and diversity.
	<li>graphically represent the inferred tree and plots of divergence, diversity, and distance distribution histograms.
	<li>detect, visualize and numerically summarize phylogenetically informative sites as well as private mutations.
	<li>compare intra-individual genetic sequence diversity between populations.
	<li>define clusters by a genetic similarity cluster identification tool that includes two approaches (TN93 pairwise genetic distance and
	Phylogenetic branch lengths)
	</ul>
	</div>
	<div align="center">
		<img width=450 src="/static/images/DIVERfront_v2.jpg">
	</div>
	<div>
	<p>Please cite the following paper if you use DIVEIN:</p> 
	<dl>
	<dd>Deng W, Maust BS, Nickle DC, Learn GH, Liu Y, Heath L, Kosakovsky Pond SL, Mullins JI 2010 
	DIVEIN: a web server to analyze phylogenies, sequence divergence, diversity, and informative sites. BioTechniques 48:405-408
	(<a href=https://www.future-science.com/doi/pdf/10.2144/000113370>PDF</a>).
	</dl>
	</div>
	</div>
	<div id="footer" align="center">
		<p class="copyright">&copy; 2025 Fred Hutch Cancer Center. All rights reserved.</p>
	</div>	
	<!-- mail server message goes here -->	
	<!-- end of message -->
</body>
</html>
